### PR TITLE
Proof of concept for PositionOrUnit

### PR DESCRIPTION
**Pull request by SonkoMagnus**
I really don't like the BWMirror implementation of this, so added a less error-prone, and better structured (IMHO) one.

### CODE REVIEW

**Comment by adakitesystems** [comment]
Hi, @SonkoMagnus!

First of all, thank you for the PR. However, I am not a fan of this enum. I would rather use `instanceof` in this case to determine its type. On the contrary, an enum for type may reduce overhead albeit minutely. 

As a minor nitpick, I am not a fan of this block:
```java
    if (target.getPOType() == PositionOrUnit.PositionOrUnitType.UNIT) {
      unitCommand.unitCommandType = UnitCommandType.Attack_Unit;
    }
    if (target.getPOType() == PositionOrUnit.PositionOrUnitType.POSITION) {
      unitCommand.unitCommandType = UnitCommandType.Attack_Move;
    }
    if (target.getPOType() == PositionOrUnit.PositionOrUnitType.TILE_POS) {
      unitCommand.unitCommandType = UnitCommandType.Attack_Move;
    }
    if (target.getPOType() == PositionOrUnit.PositionOrUnitType.WALK_POS) {
      unitCommand.unitCommandType = UnitCommandType.Attack_Move;
      ;
    }
```
...it may become error-prone and redundant when used in other methods when manually typing. Although, an easy fix would probably be to move this to a static method to return the proper `unitCommandType`. So, that's not much of an issue. 

In summary, I like the overall idea, but not really the implementation too much. However, I don't have any better ideas other than keeping it the same as in BWMirror.

Pinging @IMP-42, @dgant, @Bytekeeper, and anyone else for their opinions.

Examples for this context:

https://github.com/OpenBW/BWAPI4J/blob/017e819fca1f5e1358aab00d6dc8db59810cba22/BWAPI4J/src/main/java/org/openbw/bwapi4j/UnitCommand.java#L60-L62

https://github.com/vjurenka/BWMirror/blob/8f8882f7b0166ba73a5e33a48e241231926fa758/api/bwapi/UnitCommand.java#L6:
```java
public static native UnitCommand attack(Unit unit, PositionOrUnit target);
```

**Comment by SonkoMagnus** [comment]
As discussed, i considered instanceof - it's as good as this, might be better, in fact. Also, we can extract the logic to a static method, i agree. Like it says, it's just a POC - i'm fully agreeable on the implementation details :) Curious about thet others' opinions.

**Comment by Bytekeeper** [comment]
Java allows method overloading, that `instanceof` block could just be 4-5 attack methods.
The new `Type` also is more or less a placeholder of the different classes. It looks a bit like `C` in `Java` to me :)

**Comment by adakitesystems** [comment]
> Java allows method overloading, that instanceof block could just be 4-5 attack methods.

True but 4-5 methods is not the end of it. There is "move", "patrol", "useTech", etc. potentially 15-20 extra methods.

**Comment by Bytekeeper** [comment]
True, but the amount of code is the same :) 15-20 methods or 5 methods with 5 `instanceof`s.... Maybe we find another way, but `instanceof` is not the way to go.

**Comment by SonkoMagnus** [comment]
@Bytekeeper Or the 5 methods can call just one to determine the type. And `instanceof` is supposedly the fastest way to compare subclasses.  What do you think of the general concept? It preserves the BWMirror usage.

**Comment by adakitesystems** [comment]
This PR has sparked good discussion about `AbstractPoint` and `PositionOrUnit`. It seems the general consensus is to exclude these classes from the raw interface. I have already deleted `PositionOrUnit` and will not implement `AbstractPoint`.

Thanks for the PR, @SonkoMagnus! It helped brainstorm and move decisions along.
